sklearn/decomposition/_dict_learning.py

```diff
@@ -2381,60 +2128,52 @@ def fit(self, X, y=None):
-
-            i = -1  # to allow max_iter = 0
-
-            for i, batch in zip(range(n_steps), batches):
-                X_batch = X_train[batch]
-
-                batch_cost = self._minibatch_step(
-                    X_batch, dictionary, self._random_state, i
-                )
-
-                if self._check_convergence(
-                    X_batch, batch_cost, dictionary, old_dict, n_samples, i, n_steps
-                ):
-                    break
+        # TODO(1.6): remove in 1.6
+        if self.max_iter is None:
+            warn(
+                (
+                    "`max_iter=None` is deprecated in version 1.4 and will be removed"
+                    " in version 1.6. Use the default value (i.e. `1_000`) instead."
+                ),
+                FutureWarning,
+            )
+            max_iter = 1_000
+        else:
+            max_iter = self.max_iter
 
-                # XXX callback param added for backward compat in #18975 but a common
-                # unified callback API should be preferred
-                if self.callback is not None:
-                    self.callback(locals())
+        # Attributes to monitor the convergence
+        self._ewa_cost = None
+        self._ewa_cost_min = None
+        self._no_improvement = 0
 
-                old_dict[:] = dictionary
+        batches = gen_batches(n_samples, self._batch_size)
+        batches = itertools.cycle(batches)
+        n_steps_per_iter = int(np.ceil(n_samples / self._batch_size))
+        n_steps = max_iter * n_steps_per_iter
 
-            self.n_steps_ = i + 1
-            self.n_iter_ = np.ceil(self.n_steps_ / n_steps_per_iter)
-        else:
-            # TODO remove this branch in 1.4
-            n_iter = 1000 if self.n_iter == "deprecated" else self.n_iter
+        i = -1  # to allow max_iter = 0
 
-            batches = gen_batches(n_samples, self._batch_size)
-            batches = itertools.cycle(batches)
+        for i, batch in zip(range(n_steps), batches):
+            X_batch = X_train[batch]
 
-            for i, batch in zip(range(n_iter), batches):
-                self._minibatch_step(X_train[batch], dictionary, self._random_state, i)
+            batch_cost = self._minibatch_step(
+                X_batch, dictionary, self._random_state, i
+            )
 
-                trigger_verbose = self.verbose and i % ceil(100.0 / self.verbose) == 0
-                if self.verbose > 10 or trigger_verbose:
-                    print(f"{i} batches processed.")
+            if self._check_convergence(
+                X_batch, batch_cost, dictionary, old_dict, n_samples, i, n_steps
+            ):
+                break
 
-                if self.callback is not None:
-                    self.callback(locals())
+            # XXX callback param added for backward compat in #18975 but a common
+            # unified callback API should be preferred
+            if self.callback is not None:
+                self.callback(locals())
```

Comment:
This was probably not covered in the past as well.

Reply:
indeed :)